### PR TITLE
Retire DVec class in favour of c++20 style span for device memory.

### PR DESCRIPTION
**Pull request by RAMitchell**
Both classes serve the same purpose - a non-owning memory container. Span is more generic and better designed.

### CODE REVIEW

**Comment by trivialfis** [comment]
@RAMitchell I haven't looked this in detail yet.  Glancing through the changes it seems to make sure we access the right pointer for each device there are two conditions:

- You only use it in `DeviceShard` which can keep device id as part of it.
- Remember calling `cudaSetDevice` before doing anything.

I have been reluctant to make the change since the above two conditions are hard to guarantee, and the bugs are hard to fix ... Could you explain why do you think it's solved in this PR? I'm just want to feel a little bit assurance.

**Comment by RAMitchell** [comment]
I think the above problem still exists with the DVec class - we were not calling cudaSetDevice() on every method call of DVec even when we did have the device index information available. 

This problem of device selection might be easier to approach by enforcing that each device shard has its respective device selected every time it calls a method, instead of trying to explicitly associate memory with a device.

File: src/linear/updater_gpu_coordinate.cu [hunk 1]
```diff
@@ -87,8 +87,10 @@ class DeviceShard {
 
   void UpdateGpair(const std::vector<GradientPair> &host_gpair,
                    const gbm::GBLinearModelParam &model_param) {
-    gpair_.copy(host_gpair.begin() + ridx_begin_ * model_param.num_output_group,
-                host_gpair.begin() + ridx_end_ * model_param.num_output_group);
+    dh::safe_cuda(cudaMemcpy(
```
Comment (trivialfis) [inline src/linear/updater_gpu_coordinate.cu]: async?

File: src/tree/updater_gpu_hist.cu [hunk 2]
```diff
@@ -727,13 +728,16 @@ struct DeviceShard {
       dh::safe_cuda(cudaMemsetAsync(left_counts.data().get(), 0,
                                     sizeof(int64_t) * left_counts.size()));
     }
-    thrust::sequence(ridx.CurrentDVec().tbegin(), ridx.CurrentDVec().tend());
+    thrust::sequence(
+        thrust::device_pointer_cast(ridx.CurrentSpan().data()),
```
Comment (trivialfis) [inline src/tree/updater_gpu_hist.cu]: I thought any iterator would do,  why creating a wrapper?

Comment (RAMitchell) [reply src/tree/updater_gpu_hist.cu]: If you pass a raw pointer thrust thinks the memory is on the host and crashes.

**Pull request merged**